Fail when renaming with different case on case insensitive file system

I'm using rdiff-backup 1.2.8 to backup files on a Windows machine with cygwin. The source is the local hard drive and the backup destination is an external hard drive, both NTFS. Renaming a folder called `archive` to `Archive` causes an AssertError `assert not incrp.lstat(), incrp`. This has been reported before: https://www.backupcentral.com/phpBB2/two-way-mirrors-of-external-mailing-lists-3/rdiff-backup-23/case-insensitive-filesystems-assert-not-incrp-lstat-100362/

Renaming the folder to its original name works after running `rdiff-backup --check-destination-dir "$dst"`.

I wrote a script to reproduce that error:

``` sh
#!/bin/sh

src="$(mktemp -d)"
dst="$(mktemp -d)"
echo "backup source: $src"
echo "backup destination: $dst"

mkdir "$src/archive"
rdiff-backup "$src" "$dst"

sleep 1
mv "$src/archive" "$src/Archive-tmp"
mv "$src/Archive-tmp" "$src/Archive"
rdiff-backup "$src" "$dst"

ls -la "$src"
ls -la "$dst"
rm "$src" -rf
rm "$dst" -rf
```

I ran that script on my linux machine with a FAT32 usb drive: `TMPDIR=/media/C808-E853/ sh rdiff-backup-23.sh`

It gave me a different error, but the bug is probably related:

```
backup source: /media/C808-E853/tmp.a0Op9QfBAo
backup destination: /media/C808-E853/tmp.l1QgeTHVzJ
Warning: hard linking not supported by filesystem at /media/C808-E853/tmp.l1QgeTHVzJ/rdiff-backup-data
Warning: hard linking not supported by filesystem at /media/C808-E853/tmp.l1QgeTHVzJ/rdiff-backup-data
Exception '[Errno 17] File exists: '/media/C808-E853/tmp.l1QgeTHVzJ/Archive'' raised of class '<type 'exceptions.OSError'>':
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 304, in error_check_Main
    try: Main(arglist)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 324, in Main
    take_action(rps)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 280, in take_action
    elif action == "backup": Backup(rps[0], rps[1])
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 343, in Backup
    backup.Mirror_and_increment(rpin, rpout, incdir)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/backup.py", line 51, in Mirror_and_increment
    DestS.patch_and_increment(dest_rpath, source_diffiter, inc_rpath)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/backup.py", line 243, in patch_and_increment
    ITR(diff.index, diff)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/rorpiter.py", line 284, in __call__
    branch.start_process(*args)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/backup.py", line 725, in start_process
    self.prepare_dir(diff_rorp, self.base_rp)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/backup.py", line 654, in prepare_dir
    base_rp.mkdir()
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/rpath.py", line 983, in mkdir
    self.conn.os.mkdir(self.path)

Traceback (most recent call last):
  File "/usr/bin/rdiff-backup", line 30, in <module>
    rdiff_backup.Main.error_check_Main(sys.argv[1:])
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 304, in error_check_Main
    try: Main(arglist)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 324, in Main
    take_action(rps)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 280, in take_action
    elif action == "backup": Backup(rps[0], rps[1])
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 343, in Backup
    backup.Mirror_and_increment(rpin, rpout, incdir)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/backup.py", line 51, in Mirror_and_increment
    DestS.patch_and_increment(dest_rpath, source_diffiter, inc_rpath)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/backup.py", line 243, in patch_and_increment
    ITR(diff.index, diff)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/rorpiter.py", line 284, in __call__
    branch.start_process(*args)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/backup.py", line 725, in start_process
    self.prepare_dir(diff_rorp, self.base_rp)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/backup.py", line 654, in prepare_dir
    base_rp.mkdir()
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/rpath.py", line 983, in mkdir
    self.conn.os.mkdir(self.path)
OSError: [Errno 17] File exists: '/media/C808-E853/tmp.l1QgeTHVzJ/Archive'
```

Hi @ottok!

Thank you for following up. I just did the test again. The latest development version doesn't have instructions for Windows where I encountered the problem but your latest release seems to be up-to-date (7 days old). I downloaded rdiff-backup-1.9.1b0.win32exe.zip from the "3rd beta release before 2.0.0". Is that a 2.0 preview or how do I test this?

Anyway, the tested 1.9.1b0 still encounters the same error. Here is how I reproduced it:

```sh
mkdir s
mkdir d
mkdir d/test
touch d/test/a
./rdiff-backup.exe s d
mv s/test s/Test
./rdiff-backup.exe s d
```

I got the following output:
```
$ ./rdiff-backup.exe s d
Exception 'QuotedRPath: Path: 
d/rdiff-backup-data/increments/test.2020-03-01T09-49-59Z.dir
Index: ('test.2020-03-01T09-49-59Z.dir',)
Data: {'type': 'reg', 'size': 0, 'perms': 438, 'uid': 0, 'gid': 0, 
'inode': 4503599627465515, 'devloc': 1519922045, 'nlink': 1, 'mtime': 
1583055834, 'atime': 1583056225, 'ctime': 1583056225, 'uname': None, 
'gname': None, 'win_acl': b'# file: test.2020-03-01T09-49-59Z.dir
\nO:S-1-5-21-233083988-539155894-3873057244-1002G:S-1-5-21-233083988-539155894-3873057244-513D:(A;ID;FA;;;S-1-5-21-233083988-539155894-3873057244-1002)(A;ID;0x1200a9;;;S-1-5-21-233083988-539155894-3873057244-513)(A;ID;0x1200a9;;;WD)\n'}' 
raised of class '<class 'Assertio
nError'>':
   File "rdiff_backup\Main.py", line 390, in error_check_Main
Traceback (most recent call last):
   File "rdiff-backup", line 32, in <module>
   File "rdiff_backup\Main.py", line 390, in error_check_Main
   File "rdiff_backup\Main.py", line 412, in Main
   File "rdiff_backup\Main.py", line 348, in take_action
   File "rdiff_backup\Main.py", line 434, in Backup
   File "rdiff_backup\backup.py", line 55, in Mirror_and_increment
   File "rdiff_backup\backup.py", line 281, in patch_and_increment
   File "rdiff_backup\rorpiter.py", line 316, in __call__
   File "rdiff_backup\backup.py", line 804, in start_process
   File "rdiff_backup\increment.py", line 43, in Increment
   File "rdiff_backup\increment.py", line 118, in makedir
   File "rdiff_backup\increment.py", line 143, in get_inc
AssertionError: QuotedRPath: Path: 
d/rdiff-backup-data/increments/test.2020-03-01T09-49-59Z.dir
Index: ('test.2020-03-01T09-49-59Z.dir',)
Data: {'type': 'reg', 'size': 0, 'perms': 438, 'uid': 0, 'gid': 0, 
'inode': 4503599627465515, 'devloc': 1519922045, 'nlink': 1, 'mtime': 
1583055834, 'atime': 1583056225, 'ctime': 1583056225, 'uname': None, 
'gname': None, 'win_acl': b'# file: test.2020-03-01T09-49-59Z.dir
\nO:S-1-5-21-233083988-539155894-3873057244-1002G:S-1-5-21-233083988-539155894-3873057244-513D:(A;ID;FA;;;S-1-5-21-233083988-539155894-3873057244-1002)(A;ID;0x1200a9;;;S-1-5-21-233083988-539155894-3873057244-513)(A;ID;0x1200a9;;;WD)\n'}
[6380] Failed to execute script rdiff-backup
```

Hello I encountered the same error on Windows 10 using rdiff-backup 2.0.3. I encountered the same error as mkllnk by renaming a folder:
```
>mkdir from
>mkdir from\test
>mkdir to
>rdiff-backup.exe from to
>rename from\test Test
>rdiff-backup.exe from to
>rdiff-backup.exe from to
Exception 'QuotedRPath: Path: to/rdiff-backup-data/increments/test.2020-06-07T14-37-23+02-00.dir
Index: ('test.2020-06-07T14-37-23+02-00.dir',)
Data: {'type': 'reg', 'size': 0, 'perms': 438, 'uid': 0, 'gid': 0, 'inode': 26458647811148361, 'devloc': 2118428268, 'nlink': 1, 'mtime': 1591533431, 'atime': 1591533501, 'ctime': 1591533501, 'uname': None, 'gname': None, 'win_acl': b'# file: test.2020-06-07T14-37-23+02-00.dir\nO:BAG:S-1-5-21-2341865278-2280168835-870530268-513D:AI(A;ID;FA;;;SY)(A;ID;FA;;;BA)(A;ID;FA;;;S-1-5-21-2341865278-2280168835-870530268-1001)\n'}' raised of class '<class 'AssertionError'>':
  File "rdiff_backup\Main.py", line 393, in error_check_Main
  File "rdiff_backup\Main.py", line 415, in Main
  File "rdiff_backup\Main.py", line 351, in take_action
  File "rdiff_backup\Main.py", line 437, in Backup
  File "rdiff_backup\backup.py", line 55, in Mirror_and_increment
  File "rdiff_backup\backup.py", line 281, in patch_and_increment
  File "rdiff_backup\rorpiter.py", line 316, in __call__
  File "rdiff_backup\backup.py", line 820, in start_process
  File "rdiff_backup\increment.py", line 43, in Increment
  File "rdiff_backup\increment.py", line 118, in makedir
  File "rdiff_backup\increment.py", line 143, in get_inc

Traceback (most recent call last):
  File "rdiff-backup", line 32, in <module>
  File "rdiff_backup\Main.py", line 393, in error_check_Main
  File "rdiff_backup\Main.py", line 415, in Main
  File "rdiff_backup\Main.py", line 351, in take_action
  File "rdiff_backup\Main.py", line 437, in Backup
  File "rdiff_backup\backup.py", line 55, in Mirror_and_increment
  File "rdiff_backup\backup.py", line 281, in patch_and_increment
  File "rdiff_backup\rorpiter.py", line 316, in __call__
  File "rdiff_backup\backup.py", line 820, in start_process
  File "rdiff_backup\increment.py", line 43, in Increment
  File "rdiff_backup\increment.py", line 118, in makedir
  File "rdiff_backup\increment.py", line 143, in get_inc
AssertionError: QuotedRPath: Path: to/rdiff-backup-data/increments/test.2020-06-07T14-37-23+02-00.dir
Index: ('test.2020-06-07T14-37-23+02-00.dir',)
Data: {'type': 'reg', 'size': 0, 'perms': 438, 'uid': 0, 'gid': 0, 'inode': 26458647811148361, 'devloc': 2118428268, 'nlink': 1, 'mtime': 1591533431, 'atime': 1591533501, 'ctime': 1591533501, 'uname': None, 'gname': None, 'win_acl': b'# file: test.2020-06-07T14-37-23+02-00.dir\nO:BAG:S-1-5-21-2341865278-2280168835-870530268-513D:AI(A;ID;FA;;;SY)(A;ID;FA;;;BA)(A;ID;FA;;;S-1-5-21-2341865278-2280168835-870530268-1001)\n'}
[12408] Failed to execute script rdiff-backup
```

I also tested the issue by renaming a file instead of a folder. This resulted in a different issue which can lead to data loss. Renaming a file to the same name with another case and backup up again will delete the file from the backup:
```
>mkdir from
>mkdir to
>echo 1>from/test.txt
>rdiff-backup.exe from to
>ren from\test.txt Test.txt
>rdiff-backup.exe from to
>rdiff-backup.exe --list-increments to
Found 1 increments:
    increments.2020-06-07T14-31-12+02-00.dir   Sun Jun  7 14:31:12 2020
Current mirror: Sun Jun  7 14:32:01 2020

>rdiff-backup --list-changed-since 5D to
changed .
new     Test.txt
deleted test.txt

>dir from
 Directory of C:\Users\Mat\Downloads\rdiff-backup-2.0.3\from

07-06-2020  14:31    <DIR>          .
07-06-2020  14:31    <DIR>          ..
07-06-2020  14:30                13 Test.txt
               1 File(s)             13 bytes
               2 Dir(s)  48.765.251.584 bytes free

>dir to
 Directory of C:\Users\Mat\Downloads\rdiff-backup-2.0.3\to

07-06-2020  14:31    <DIR>          .
07-06-2020  14:31    <DIR>          ..
07-06-2020  14:32    <DIR>          rdiff-backup-data
               0 File(s)              0 bytes
               3 Dir(s)  48.765.251.584 bytes free
```

--- a/docs/FAQ.adoc
+++ b/docs/FAQ.adoc
@@ -373,3 +373,18 @@ The list might grow but issues are currently known with especially slow combinat
 
 * Fuse-exFAT (the newer kernel version seems fine though)
 * "complex" file systems on a NAS (e.g. ZFS on SMB or NFS)
+* SSHFS
+
+== How to backup safely from case aware to case aware file system?
+
+File systems like VFAT or NTFS are what I call case aware file systems:
+they are "officially" case insensitive but they store file names in a case sensitive manner.
+
+This becomes a https://github.com/rdiff-backup/rdiff-backup/issues/11[problem for rdiff-backup if a file is renamed in a case sensitive manner] (e.g. from `MyFile` to `mYfIlE`) because it doesn't recognize them as the same file, but the target file system doesn't accept both files next to each other.
+
+You can either:
+
+. make sure that files are not renamed in such a manner (this doesn't happen that often)
+. or you backup _from the beginning_ using the option `--override-chars-to-quote 'A-Z-"*/:<>?\\|;'`, which makes sure that capital letters (and other "strange" characters) are quoted so that there is no name collision on the file system.
++
+CAUTION: remember that you can't change the quoting once a backup repository has been setup, so that you might need to create a new one.

